File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-2.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1

        # Find the start of the sorted array
        while l < r:
            m = (r + l) // 2
            if nums[m] > nums[r]:
                l = m + 1
            else:
                r = m

        offset = l
        
        l, r = 0, len(nums) - 1
        while l <= r:
            m = (r + l) // 2
            offset_m = m + offset
            if offset_m >= len(nums):
                offset_m -= len(nums)
            print(l, r, m, offset_m)
            if nums[offset_m] == target:
                return offset_m
            elif nums[offset_m] > target:
                r = m - 1
            else:
                l = m + 1
            print(l, r, m, offset_m)
            print()
            

        return -1

File: Data Structures & Algorithms/binary-search/submission-2.java
class Solution {
    public int search(int[] nums, int target) {
        return searchHelper(nums, target, nums.length / 2, 0, nums.length - 1);
    }

    public int searchHelper(int[] nums, int target, int middle, int start, int end) {
        int l = 0, r = nums.length - 1;

        while (l <= r) {
            int m = l + ((r - l) / 2);
            if (nums[m] > target) {
                r = m - 1;
            } else if (nums[m] < target) {
                l = m + 1;
            } else {
                return m;
            }
        }
        return -1;
    }
}
